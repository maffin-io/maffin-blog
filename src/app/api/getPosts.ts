import { DateTime } from 'luxon';
import { Octokit } from 'octokit';

import markdownToHtml from '@/lib/markdownToHtml';

export type Post = {
  slug: string,
  date: DateTime,
  title: string,
  summary: string,
  reading_time: string,
  tags: string[],
  content: string,
  author: {
    name?: string,
    avatar?: string,
  },
};

export async function getPosts(): Promise<Post[]> {
  const octokit = new Octokit();
  const issues = (await octokit.rest.issues.listForRepo({
    owner: 'maffin-io',
    repo: 'maffin-blog',
    creator: 'argaen',
    headers: {
      accept: 'application/vnd.github+json',
    },
  })).data.filter(issue => !issue.pull_request && !issue.title.includes('[draft]'));

  return Promise.all(issues.map(async (issue) => {
    if (!issue.body) {
      throw new Error(`Missing body in issue ${issue.title}`);
    }

    const { content: htmlContent, metadata } = await markdownToHtml(issue.body);
    return {
      slug: metadata.slug,
      date: DateTime.fromISO(issue.created_at),
      title: issue.title,
      summary: metadata.summary,
      reading_time: metadata.reading_time,
      // @ts-ignore
      tags: issue.labels.map(label => label.name),
      content: htmlContent,
      author: {
        name: issue.user?.name || issue.user?.login,
        avatar: issue.user?.avatar_url,
      },
    };
  }));
}

export async function getPost(slug: string): Promise<null | Post> {
  const posts = await getPosts();
  return posts.find(post => post.slug === slug) || null;
}

export async function getPostsByTag(tag: string): Promise<Post[]> {
  const posts = await getPosts();
  return posts.filter(post => post.tags.includes(tag));
}
