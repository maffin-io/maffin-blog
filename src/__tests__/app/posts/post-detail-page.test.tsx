import { DateTime } from 'luxon';
import { render } from '@testing-library/react';

import BlogPage from '@/app/posts/[slug]/page';
import * as postsApi from '@/app/api/getPosts';

jest.mock('@/app/api/getPosts', () => ({
  __esModule: true,
  ...jest.requireActual('@/app/api/getPosts'),
}));

/**
 * @param {function} Component
 * @param {*} props
 * @returns {Promise<()=>JSX.Element>}
 */
async function resolveComponent(Component: React.FunctionComponent, props: any) {
  const ComponentResolved = await Component(props);
  return () => ComponentResolved;
}

describe('BlogPage', () => {
  beforeEach(() => {
    jest.spyOn(postsApi, 'getPost').mockResolvedValue(null);
  });

  it('renders post with content and title', async () => {
    jest.spyOn(postsApi, 'getPost').mockResolvedValue({
      content: '<div>Content</div>',
      slug: 'my-blog-post',
      summary: 'Blog post summary TODO',
      tags: ['label1'],
      reading_time: '1 min',
      title: 'My Blog Post',
      date: DateTime.fromISO('2023-01-01'),
      author: {
        name: 'Name',
        avatar: 'https://avatar.image',
      },
    });
    const Component = await resolveComponent(BlogPage, { params: { slug: 'My-Blog-Post' } });
    const { container } = render(<Component />);

    expect(container).toMatchSnapshot();
  });

  it('returns 404 when slug not found', async () => {
    await expect(
      () => resolveComponent(BlogPage, { params: { slug: 'My-Blog-Post%3A-Part-1' } }),
    ).rejects.toThrow('NEXT_NOT_FOUND');
  });
});
