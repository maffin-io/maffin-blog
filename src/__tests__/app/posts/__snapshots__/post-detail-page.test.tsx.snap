// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogPage renders post with content and title 1`] = `
<div>
  <article>
    <div
      class="xl:divide-y xl:divide-gray-200 xl:dark:divide-gray-700"
    >
      <header
        class="pt-6 xl:pb-6"
      >
        <div
          class="space-y-1 text-center"
        >
          <dl
            class="space-y-10"
          >
            <div>
              <dt
                class="sr-only"
              >
                Published on
              </dt>
              <dd
                class="text-base font-medium leading-6 text-gray-500 dark:text-gray-400"
              >
                <time
                  datetime="2023-01-01"
                >
                  Jan 1, 2023
                </time>
              </dd>
            </div>
          </dl>
          <div>
            <h1
              class="text-3xl font-extrabold leading-9 tracking-tight text-gray-900 dark:text-gray-100 sm:text-4xl sm:leading-10 md:text-5xl md:leading-14"
            >
              My Blog Post
            </h1>
          </div>
        </div>
      </header>
      <div
        class="divide-y divide-gray-200 pb-8 dark:divide-gray-700 xl:grid xl:grid-cols-4 xl:gap-x-6 xl:divide-y-0"
        style="grid-template-rows: auto 1fr;"
      >
        <dl
          class="pt-6 pb-10 xl:border-b xl:border-gray-200 xl:pt-11 xl:dark:border-gray-700"
        >
          <dt
            class="sr-only"
          >
            Author
          </dt>
          <dd>
            <ul
              class="flex justify-center space-x-8 sm:space-x-12 xl:block xl:space-x-0 xl:space-y-8"
            >
              <li
                class="flex items-center space-x-2"
              >
                <img
                  alt="avatar"
                  class="h-10 w-10 rounded-full"
                  data-nimg="1"
                  decoding="async"
                  height="38"
                  loading="lazy"
                  src="/_next/image?url=https%3A%2F%2Favatar.image&w=96&q=75"
                  srcset="/_next/image?url=https%3A%2F%2Favatar.image&w=48&q=75 1x, /_next/image?url=https%3A%2F%2Favatar.image&w=96&q=75 2x"
                  style="color: transparent;"
                  width="38"
                />
                <dl
                  class="whitespace-nowrap text-sm font-medium leading-5"
                >
                  <dt
                    class="sr-only"
                  >
                    Name
                  </dt>
                  <dd
                    class="text-gray-900 dark:text-gray-100"
                  >
                    Name
                  </dd>
                  <dt
                    class="sr-only"
                  >
                    Twitter
                  </dt>
                </dl>
              </li>
            </ul>
          </dd>
        </dl>
        <div
          class="divide-y divide-gray-200 dark:divide-gray-700 xl:col-span-3 xl:row-span-2 xl:pb-0"
        >
          <div
            class="prose prose-img:mx-auto max-w-none pt-10 pb-8 dark:prose-dark"
          >
            <div>
              <div>
                Content
              </div>
            </div>
          </div>
        </div>
        <footer>
          <div
            class="divide-gray-200 text-sm font-medium leading-5 dark:divide-gray-700 xl:col-start-1 xl:row-start-2 xl:divide-y"
          >
            <div
              class="py-4 xl:py-8"
            >
              <h2
                class="text-xs uppercase tracking-wide text-gray-500 dark:text-gray-400"
              >
                Tags
              </h2>
              <div
                class="flex flex-wrap"
              >
                <a
                  href="/tags/label1"
                >
                  <span
                    class="mr-3 text-sm font-medium uppercase text-primary-500 hover:text-primary-600 dark:hover:text-primary-400"
                  >
                    label1
                  </span>
                </a>
              </div>
            </div>
          </div>
          <div
            class="divide-gray-200 font-medium leading-5 dark:divide-gray-700 xl:col-start-1 xl:row-start-2 xl:divide-y"
          >
            <div
              class="pt-4 xl:pt-8"
            >
              <a
                class="text-primary-500 hover:text-primary-600 dark:hover:text-primary-400"
                href="/"
              >
                ‚Üê Back to the blog
              </a>
            </div>
          </div>
        </footer>
      </div>
    </div>
  </article>
</div>
`;
